keps/sig-api-machinery/20190621-apply-scalability.md

```diff
@@ -0,0 +1,154 @@
+
+The internal format could be stored in protobuf, which would make it much smaller than json (json has a lot of repeated fields that make it extremely large).
+
+### Only update managedFields if previously applied to
+
+Most objects created with post or put are never applied to, so tracking managedFields on them is a waste of resources. If an object without managedFields does get applied to, we can still generate a single managedFields entry for all the existing fields. The generated managedFields conflicts set will be the same, but we won’t have any way of knowing know who is being conflicted with. CPU and memory cost on the apiserver will be dependent on the number of objects managed by server-side apply. It would increase more gradually as controllers decided to adopt SSA. Adoption could be decided on a controller by controller basis, weighing the cost vs. benefits each time. If someone applies to an object created by post or put, the conflicts they see would not contain information about the timestamp or field manager, so conflicts become less useful.
+
+### Storing in one format, serving in another
+
+Another possible solution would be to combine many of these solutions:
+
+Some combination of storing the field managers as one big blob in etcd, using the internal fieldpath representation, as gzipped protobuf, would give us a much lower overhead in etcd and very fast serialization/deserialization. This is especially true if deserialization doesn’t need to know about the specifics of managedfields (hence just deserializing as a blob).
+
+Serving the API as it currently is (or maybe changing it if we believe we need to change what we serve), but only on demand so that the price of deserializing would only be paid by clients that specifically request it.
+
+Based on initial numbers, this solution has the most advantages:
+etcd objects (based on pods above) only grow by 30%
+listing objects has nearly zero cost since we don’t send unless requested,
+patches and updates can deserialize the data very quickly directly in protobuf, and process the data
+API is still “usable”
+
+This could be implemented as a subresource similar to scale. (something like “/api/v1/pods/managedfields”) The subresource would decode the stored form of managedFields and exposes an object of type meta/v1.ManagedFields which would be highly structured. The form would be similar to or the same as how managedFields is stored today. This part could be implemented at a future date.
```

Comment:
This seems like it would have atomicity issues if a  user wanted to claim ownership. Wasn't there a use-case where we manually twiddled the field ownership?

Reply:
That's true, this would remove the ability to easily tie a specific managed fields to a specific version of the object, unless resource versions were used, or the managedfields sub resource  could serve the entire object + the managedfields